### DISCUSSION TITLE
Wante to change the current context to my custom one.

### DISCUSSION THREAD

**Discussion by klaiderman** [python-telegram-bot/python-telegram-bot#3948] [Q&A] [answered]
I made a custom command decorator so that it would apply to all of my commands via typing but it doesn't seem to have been so successful. I would appreciate it if someone would help me to set it up in the bot settings.
```python
class BetterContext(CallbackContext):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        
   async def send(self, ...):
       ...
```

```python
def command(name):
    def decorator(func):
        async def wrapper(_, update: Update, context: BetterContext):
            await func(_, update, context)
        wrapper.is_command = True
        return wrapper 
    return decorator
 ```
 ```python
@command(name='demo')
 async def demo_cmd(self, update: Update, context: BetterContext): # This command is part of a class
     await context.send(text='test')                               # so, the self is necessary :D
 ```
 ```cmd
Traceback (most recent call last):
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\telegram\ext\_application.py", line 1195, in process_update
    await coroutine
  File "C:\Users\User\AppData\Local\Programs\Python\Python39\lib\site-packages\telegram\ext\_basehandler.py", line 153, in handle_update
    return await self.callback(update, context)
  File "C:\Users\User\source\repos\bot\utils.py", line 123, in wrapper
    await func(self, update, context)
  File "C:\Users\User\source\repos\bot\bot.py", line 50, in demo_cmd
    await context.send(text='test')
AttributeError: 'CallbackContext' object has no attribute 'send'
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ ^

**Comment by Bibo-Joshi** [comment, marked as answer]
Type hinting doesn't change the type of the argument that you pass. Please have a look at https://docs.python-telegram-bot.org/en/v20.6/telegram.ext.contexttypes.html and https://docs.python-telegram-bot.org/en/v20.6/examples.html#examples-contexttypesbot for info on how to make PTB use your sublass.

**Reply by klaiderman** [reply]
Well, I see. thanks!
By the way, it goes the same way if I want to customize the Update by my own?

**Reply by Bibo-Joshi** [reply]
No. PTB does not provide any ready made way to customize that class - as the contents of the `Update` class are defined by Telegram, there way way fewer use cases for that than for customizing the `CallbackContext` class.
If you use `run_polling`, you'll have to override `Bot.get_updates` and then use `ApplicationBuilder.bot()` to you a custom bot instance.
For `run_webhook`, there is currently no way to achieve this without digging into the (private) internals of PTB. You would have to use a custom webhook setup and use `YourUpdateClass.de_json` instead of `Update.de_json`.

**Reply by klaiderman** [reply]
I see, I have asked this because I want to get from the `CallbackContext` object (my `BetterContext` object) those `effective_user`, `effective_message`, and `effective_chat` attributes. But, I have no access to them from the context object. I have tried to find `CallableContext`'s methods that would transfer this data for me, but I haven't found it so far. Tried this trick as-well, but no succession,
```py
def command(name):
    def decorator(func):
        async def wrapper(self, update: Update, context: BetterContext):
            await func(self, update, context)
            for attr in ('effective_chat', 'effective_user', 'effective_message'):
                value = getattr(update, attr)
                setattr(context, attr, value)
        wrapper.is_command = True
        return wrapper 
    return decorator

**Reply by Bibo-Joshi** [reply]
PTB makes a clear distinction between the `update` argument and the `context` argument to emphasize which information orginates from Telegram (the `update`) and what is PTB-specific infrastructure (`context`). the `effective_*` attributes are pure shortcuts to information within the `update` argument and are hence located in the `Update` class. I see no added benefit in having them available in `context` as well.

In any case, one way to make that work is to extract these values from `update` in `BetterContext.from_update` (which would override `CallbackContext.from_update`, as shocased in the examples linked above).
